kornia/geometry/transform/image_registrator.py

```diff
@@ -0,0 +1,263 @@
+        optimizer: optimizer class used for the optimization
+        loss_fn: torch loss function,
+        pyramid_levels (int): number of scale pyramid levels,
+        lr(float): learning rate for optimization,
+        num_iterations(int): maximum number of iterations,
+        tolerance(float): stop optimizing if loss difference is less. default 1e-4,
+        warper: if model_type is not string, one needs to provide warper object"""
+    known_models = ['homography',
+                    'similarity',
+                    'translation',
+                    'scale',
+                    'rotation']
+
+    def __init__(self,
+                 model_type='homography',
+                 optimizer=optim.Adam,
+                 loss_fn=F.l1_loss,
+                 pyramid_levels: int = 5,
+                 lr: float = 1e-3,
+                 num_iterations: int = 100,
+                 tolerance: float = 1e-4,
+                 warper=None):
+        super().__init__()
+        # We provide pre-defined combinations or allow user to supply model
+        # together with warper
+        if type(model_type) is not str:
+            if warper is None:
+                raise ValueError("You must supply warper together with custom model")
+            self.warper = warper
+            self.model = model_type
+        else:
+            if model_type.lower() == "homography":
+                self.warper = HomographyWarper
+                self.model = Homography()
+            elif model_type.lower() == "similarity":
+                self.warper = HomographyWarper
+                self.model = Similarity(True, True, True)
+            elif model_type.lower() == "translation":
+                self.warper = HomographyWarper
+                self.model = Similarity(False, False, True)
+            elif model_type.lower() == "rotation":
+                self.warper = HomographyWarper
+                self.model = Similarity(True, False, False)
+            elif model_type.lower() == "scale":
+                self.warper = HomographyWarper
+                self.model = Similarity(False, True, False)
+            else:
+                raise ValueError(f"{model_type} is not supported. Try {self.known_models}")
+        self.pyramid_levels = pyramid_levels
+        self.optimizer = optimizer
+        self.lr = lr
+        self.loss_fn = loss_fn
+        self.num_iterations = num_iterations
+        self.tolerance = tolerance
+
+    def get_single_level_loss(self,
+                              img_src: torch.Tensor,
+                              img_dst: torch.Tensor,
+                              transform_model: torch.nn.Module) -> torch.Tensor:
+        """Warps img_src into img_dst with transform_model and returns loss."""
+        # ToDo: Make possible registration of images of different shape
+        assert len(img_src.shape) == len(img_dst.shape), (img_src.shape, img_dst.shape)
```

Comment:
raise an error with meaningful message to fix pre-commit and deepsource error message